exec/Chainweb/Gaps.hs

```diff
@@ -6,154 +6,209 @@
+        let doChain (cid, gs) = do
+              let ranges = concatMap (createRanges cid) gs
+              f logg count cid ranges
+        let gapFiller = do
+              race_ (progress logg count totalNumBlocks)
+                    (traverseConcurrently_ Par' doChain (M.toList gapsByChain))
+              final <- readIORef count
+              logg Info $ fromString $ printf "Filled in %d missing blocks." final
+        if disableIndexesPred
+          then withDroppedIndexes pool logg gapFiller
+          else gapFiller
   where
     pool = _env_dbConnPool env
+    delay =  _gapArgs_delayMicros args
+    disableIndexesPred =  _gapArgs_disableIndexes args
+    gi = mkGenesisInfo $ _env_nodeInfo env
     logg = _env_logger env
-    genesisInfo = mkGenesisInfo $ _env_nodeInfo env
-    delayFunc =
-      case delay of
-        Nothing -> pure ()
-        Just d -> threadDelay d
-    f :: LogFunctionIO Text -> IORef Int -> (BlockHeight, Int) -> IO ()
-    f logger count (h, cid) = do
-      let range = (ChainId cid, Low h, High h)
-      headersBetween env range >>= \case
-        Left e -> logger Error $ fromString $ printf "ApiError for range %s: %s" (show range) (show e)
-        Right [] -> logger Error $ fromString $ printf "headersBetween: %s" $ show range
-        Right hs -> traverse_ (writeBlock env pool count) hs
-      delayFunc
-
---queryGaps :: Env -> IO (BlockHeight, Int, Int)
---queryGaps e = P.withResource pool $ \c -> runBeamPostgres c $ do
---  gaps <- runSelectReturningList
---    $ select
---    $ do
---      (chain, height, nextHeight) <-
---        orderBy_ (asc_ . snd) $
---        aggregate_ (\(c,h) -> (group_ (c,h), lead1_ h)) $ do
---          blk <- all_ $ _cddb_blocks database
---          pure (_block_chainId blk, _block_height blk)
---      guard_ (nextHeight - height >. val_ 1)
---  pure $ NEL.nonEmpty pairs >>= filling cids . expanding . grouping
+    fi = fromIntegral
+    createRanges cid (low, high)
+      | low == high = []
+      | fromIntegral (genesisHeight (ChainId (fromIntegral cid)) gi) == low = rangeToDescGroupsOf blockHeaderRequestSize (Low $ fromIntegral low) (High $ fromIntegral (high - 1))
+      | otherwise = rangeToDescGroupsOf blockHeaderRequestSize (Low $ fromIntegral (low + 1)) (High $ fromIntegral (high - 1))
 
--- with gaps as (select chainid, height, LEAD (height,1) OVER (PARTITION BY chainid ORDER BY height ASC) AS next_height from blocks group by chainid, height) select * from gaps where next_height - height > 1;
+    f logger count cid gs =
+      S.each gs
+      & S.delay (maybe 0 (\d -> fi d / fi 1000000000) delay)
+      & S.mapM (\(l,h) -> let r = (ChainId (fromIntegral cid),l,h) in  fmap (\res -> first (\err -> (r,err)) res) $ headersBetween env r)
+      & S.partitionEithers
+      & S.chain (\(r,e) -> lift $ logger Error (fromString $ printf "ApiError for range %s" (show r) (show e)))
+      & S.effects
+      & S.concat
+      & chunksOf 1000
```

Comment:
Batching in chunks of 1000 seems to slow things down.

```
2021-08-06T18:23:05.358Z [Info] [] Progress: 440000/28726786 (1.53%), ~32 minutes remaining at 879669 items per minute.
```

vs

```
2021-08-06T18:32:43.370Z [Info] [] Progress: 505800/28728175 (1.76%), ~27 minutes remaining at 1011575 items per minute.
```

for `chunksOf 360` on my machine.